src/main/java/org/zalando/nakadi/ShutdownHooks.java

```diff
@@ -0,0 +1,50 @@
+package org.zalando.nakadi;
+
+import java.io.Closeable;
+import java.util.HashSet;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ShutdownHooks {
+
+    private static final Set<Runnable> HOOKS = new HashSet<>();
+    private static final Logger LOG = LoggerFactory.getLogger(ShutdownHooks.class);
+
+    static {
+        Runtime.getRuntime().addShutdownHook(new Thread(ShutdownHooks::onNodeShutdown));
+    }
+
+    private static void onNodeShutdown() {
+        boolean haveHooks = true;
+        while (haveHooks) {
+            final Runnable hook;
+            synchronized (HOOKS) {
+                hook = HOOKS.isEmpty() ? null : HOOKS.iterator().next();
+                HOOKS.remove(hook);
```

Comment:
I think `Queue` would be a better data structure for `HOOKS`. Above two lines you would be able to replace with `hook = HOOKS.poll();`